public/app/features/explore/state/query.ts

```diff
@@ -402,6 +438,23 @@ export const runQueries = (
             console.error(error);
           }
         );
+
+      // @ts-ignore
+      const logsVolumeDataProvider = datasourceInstance.getLogsVolumeDataProvider
+        ? // @ts-ignore
+          datasourceInstance.getLogsVolumeDataProvider(transaction.request)
+        : undefined;
+
+      dispatch(
+        storeLogsVolumeDataProviderAction({
+          exploreId,
+          logsVolumeDataProvider: logsVolumeDataProvider,
```

Comment:
why do we need to store this in redux? isn't it already available in the datasource instance itself?

more generally I don't think it's a good idea to store this in the state. i think we should find a way to access the providers based via `getDatasourceSrv().get(datasource.id)` and call them when needed, for example in `loadLogsVolumeData` thunk.

Reply:
Provider is not coupled with data source instance, but with the query. Single data source instance can provide multiple providers - for example when Explore is in split mode, each side may have two totally different providers.

Reply:
but when running the query, inside the `runQueries` thunk, you have access to the queries being run as well.

storing the provider in the topmost store slice, binds an explore pane to a single provider and given multiple datasources in explore is something we in roadmap, this will need to be refactored anyway.

the point is that I don't see any real value in having the providers in the store, given they can be obtained via other properties in the store itself, so that we can keep the implementation simpler by not caring about them.

i created a branch at https://github.com/Elfo404/grafana/tree/proposal where providers are not stored in redux, it's not really polished but it might help explaining my point.

also, maybe I am missing something else here, maybe we can have a chat and go through it together?

Reply:
As discussed offline: the provider depends on queries and time range so indeed it could be derived on runtime and not stored in the state. The only problem is that provider uses queries and time range that was used to run the query, which may be different than the current state.

We're gonna keep this functionality internal for now until it's implemented in more internal data sources and we can review it when we decide to make the API public.